File: C/HDBn/HDBn.c
// n replace -1

#include <stdio.h>
#include <stdlib.h>

#define n 3

int sourceSize = 60;

char *decode(char *coding)
{
    char *result = (char *)malloc(sourceSize * sizeof(char));
    int index = 0;
    int last = 1;
    while (coding[index] != '\0')
    {
        if (coding[index] == 'n')
        {
            result[index] = '1';
            if (last == -1)
            {
                for (int i = index; i >= index - n; i--)
                {
                    result[i] = '0';
                }
            }
            else
            {
                last = -last;
            }
        }
        else if (coding[index] == '1')
        {
            result[index] = '1';
            if (last == 1)
            {
                for (int i = index; i >= index - n; i--)
                {
                    result[i] = '0';
                }
            }
            else
            {
                last = -last;
            }
        }
        else if (coding[index] == '0')
        {
            result[index] = '0';
        }
        index++;
    }
    result[index] = '\0';
    return result;
}

char *encode(char *source)
{
    char *result = (char *)malloc(sourceSize * sizeof(char));
    int index = 0;
    int counter = 0;
    int last = 1;  // The first 1 is encoded as -1
    int lastV = 1; // The first V is encoded as -V
    while (source[index] != '\0')
    {
        if (source[index] == '1')
        {
            counter = 0; // A 1 indicates continuous 0 interruption
            result[index] = last == 1 ? 'n' : '1';
            last = -last;
        }
        else if (source[index] == '0')
        {
            counter++;
            result[index] = '0';
            if (counter == n + 1)
            {
                counter = 0;
                result[index] = lastV == 1 ? 'n' : '1';
                lastV = -lastV;
                if (result[index - n - 1] != result[index])
                {
                    result[index - n] = result[index];
                }
            }
        }
        index++;
    }
    result[index] = '\0'; // Add a terminator '\0' to the result manually
    return result;
}

char *readFile(FILE *f)
{
    char *source = (char *)malloc(sourceSize * sizeof(char));
    fgets(source, sourceSize, f);
    while (!feof(f))
    {
        rewind(f); // Re-read from the beginning of the file each time
        sourceSize = sourceSize * 1.5;
        source = realloc(source, sourceSize);
        fgets(source, sourceSize, f);
    }
    return source;
}

void main()
{
    FILE *f = fopen("test.txt", "r");
    if (f == NULL)
    {
        printf("file is not exist");
        exit(1);
    }
    char *source = readFile(f);
    char *coding = encode(source);
    char *decoding = decode(coding);
    printf("source:\t\t%s\n", source);
    printf("coding:\t\t%s\n", coding);
    printf("decoding:\t%s\n", decoding);
}

File: Java/src/HDBn/HDBn.java
package HDBn;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class HDBn {
    public static final int n = 3;

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("src/HDBn/test.txt");
        String source = readFile(f);
        String coding = encode(source);
        String decoding = decode(coding);
        System.out.println("source:\t\t" + source);
        System.out.println("coding:\t\t" + coding);
        System.out.println("decoding:\t" + decoding);
    }

    public static String readFile(File f) throws FileNotFoundException {
        Scanner sc = new Scanner(f);
        return sc.next();
    }

    public static String encode(String source) {
        StringBuilder result = new StringBuilder();
        int counter = 0;
        int last = 1;
        int lastV = 1;
        for (int i = 0; i < source.length(); i++) {
            if (source.charAt(i) == '1') {
                counter = 0;
                result.append(last == 1 ? 'n' : '1');
                last = -last;
            } else if (source.charAt(i) == '0') {
                counter++;
                result.append('0');
                if (counter == n + 1) {
                    counter = 0;
                    result.setCharAt(i, lastV == 1 ? 'n' : '1');
                    lastV = -lastV;
                    if (result.charAt(i - n - 1) != result.charAt(i)) {
                        result.setCharAt(i - n, result.charAt(i));
                    }
                }
            }
        }
        return result.toString();
    }

    public static String decode(String coding) {
        StringBuilder result = new StringBuilder();
        int last = 1;
        for (int i = 0; i < coding.length(); i++) {
            if (coding.charAt(i) == 'n') {
                result.append('1');
                if (last == -1) {
                    for (int j = i; j >= i - n; j--) {
                        result.setCharAt(j, '0');
                    }
                } else {
                    last = -last;
                }
            } else if (coding.charAt(i) == '1') {
                result.append('1');
                if (last == 1) {
                    for (int j = i; j >= i - n; j--) {
                        result.setCharAt(j, '0');
                    }
                } else {
                    last = -last;
                }
            } else if (coding.charAt(i) == '0') {
                result.append('0');
            }
        }
        return result.toString();
    }
}
